File: backend/app/auth_models.py
from __future__ import annotations

from dataclasses import dataclass

from pydantic import BaseModel, Field


@dataclass(frozen=True)
class Principal:
    username: str
    role: str
    display_name: str
    must_change_password: bool
    team_member_id: int


@dataclass(frozen=True)
class DbAuthUser:
    username: str
    password_hash: str
    role: str
    display_name: str
    team_active: bool
    must_change_password: bool
    team_member_id: int


class LoginRequest(BaseModel):
    username: str = Field(min_length=1)
    password: str = Field(min_length=1)


class RefreshRequest(BaseModel):
    refresh_token: str = Field(min_length=1)


class ChangePasswordRequest(BaseModel):
    current_password: str = Field(min_length=1)
    new_password: str = Field(min_length=8)


class AdminResetPasswordRequest(BaseModel):
    username: str = Field(min_length=1)
    new_password: str = Field(min_length=8)
    force_change: bool = True


class TokenPairResponse(BaseModel):
    access_token: str
    refresh_token: str
    token_type: str = "bearer"
    expires_in_seconds: int
    must_change_password: bool


class UserResponse(BaseModel):
    username: str
    role: str
    display_name: str
    must_change_password: bool
    team_member_id: int


class BasicMessageResponse(BaseModel):
    status: str
    message: str

File: mobile/flutter_app/lib/src/models/auth_models.dart
class LoginResult {
  const LoginResult({
    required this.accessToken,
    required this.refreshToken,
  });

  final String accessToken;
  final String refreshToken;

  factory LoginResult.fromJson(Map<String, dynamic> json) {
    return LoginResult(
      accessToken: (json['access_token'] ?? '').toString(),
      refreshToken: (json['refresh_token'] ?? '').toString(),
    );
  }
}

class AuthUser {
  const AuthUser({
    required this.username,
    required this.role,
    required this.displayName,
    required this.teamMemberId,
  });

  final String username;
  final String role;
  final String displayName;
  final int teamMemberId;

  factory AuthUser.fromJson(Map<String, dynamic> json) {
    return AuthUser(
      username: (json['username'] ?? '').toString(),
      role: (json['role'] ?? '').toString(),
      displayName: (json['display_name'] ?? '').toString(),
      teamMemberId: int.tryParse((json['team_member_id'] ?? '').toString()) ?? 0,
    );
  }
}
